### PR TITLE
[Nuclio] Refactor API Gateway class to inherit from `ModelObj`

### PR DESCRIPTION
**Pull request by TomerShor**
Align API Gateway object with the rest of the MLRun objects by inheriting from `ModelObj` class.
This organizes the API Gateway properties under `metadata` and `spec`, as well as  allows using the known methods such as `to_dict` / `to_yaml`.

### CODE REVIEW

**Review by liranbg — APPROVED** [review]

File: mlrun/runtimes/nuclio/api_gateway.py [hunk 1]
```diff
@@ -134,21 +202,26 @@ def __init__(
...
+            path=path,
+            authentication=authentication,
+            functions=None,
+            canary=canary,
+        )
+
+        self._validate(
+            functions=functions,
             canary=canary,
         )
```
Comment (rokatyy) [inline mlrun/runtimes/nuclio/api_gateway.py]: Let's move validate_function and canary to APIGatewaySpec

Comment (TomerShor) [reply mlrun/runtimes/nuclio/api_gateway.py]: makes sense 👍

File: mlrun/runtimes/nuclio/api_gateway.py [hunk 2]
```diff
@@ -134,21 +295,21 @@ def __init__(
...
         :param canary: The canary percents for the API gateway of type list[int]; for instance: [20,80]
         """
-        self.functions = None
-        self._validate(
+        self.metadata = APIGatewayMetadata(
+            name=name,
             project=project,
+        )
+
+        self.spec = APIGatewaySpec(
```
Comment (alonmr) [inline mlrun/runtimes/nuclio/api_gateway.py]: see the following example on how to define meta and spec. constructor should only have meta and spec params
```
    @property
    def metadata(self) -> ProjectMetadata:
        return self._metadata

    @metadata.setter
    def metadata(self, metadata):
        self._metadata = self._verify_dict(metadata, "metadata", ProjectMetadata)

    @property
    def spec(self) -> ProjectSpec:
        return self._spec

    @spec.setter
    def spec(self, spec):
        self._spec = self._verify_dict(spec, "spec", ProjectSpec)
```

File: mlrun/runtimes/nuclio/api_gateway.py [hunk 3]
```diff
@@ -92,7 +93,167 @@ def to_scheme(
...
+            Union[
+                list[
+                    Union[
+                        RemoteRuntime,
+                        ServingRuntime,
+                    ]
+                ],
+                Union[RemoteRuntime, ServingRuntime],
+            ],
+        ],
```
Comment (alonmr) [inline mlrun/runtimes/nuclio/api_gateway.py:238]: I feel like this can be simpler. Can't union accept more than 2 options?

**Review by alonmr — CHANGES_REQUESTED** [review]

**Review by liranbg — APPROVED** [review]

**Pull request merged**